sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,328 @@ def transform(self, X):
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed from up to "n_neighbors" nearest
+    neighbors found in the training set. Each missing feature is then
+    imputed as the average, either weighted or unweighted, of these neighbors
+    who have a value for it. Where all neighbors have that feature value
+    missing, the training set average for that feature is used for imputation.
+
+    Parameters
+    ----------
+    missing_values : integer or "NaN", optional (default = "NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. For missing values encoded as np.nan,
+        use the string value "NaN".
+
+    n_neighbors : int, optional (default = 5)
+        Maximum number of neighboring samples to use for imputation. When any
+        of the neighbors themselves have the feature value missing then the
+        remaining neighbors, if any, that have the feature value available are
+        used.
+
+    weights : str or callable, optional (default = "uniform")
+        Weight function used in prediction.  Possible values:
+
+        - 'uniform' : uniform weights.  All points in each neighborhood
+          are weighted equally.
+        - 'distance' : weight points by the inverse of their distance.
+          in this case, closer neighbors of a query point will have a
+          greater influence than neighbors which are further away.
+        - [callable] : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
+    metric : str, optional (default = "masked_euclidean")
+        Distance metric for searching neighbors. Possible values:
+        - 'masked_euclidean'
+
+    row_max_missing : float, optional (default = 0.5)
+        The maximum percentage of columns (i.e. features) that can be missing
+        before the sample is excluded from nearest neighbor imputation. It
+        means that such rows will not be considered a potential donor in fit()
+        and in transform() their missing feature values will be imputed to be
+        the column mean for the entire dataset.
+
+    col_max_missing : float, optional (default = 0.8)
+        The maximum percentage of rows (or samples) that can be missing
+        for a given feature beyond which an error is raised.
+
+    copy : boolean, optional (default = True)
+        If True, a copy of X will be created. If False, imputation will
+        be done in-place whenever possible. Note that, if metric is
+        "masked_euclidean" and copy=False then missing_values in the
+        input matrix X will be overwritten with zeros.
+
+    Attributes
+    ----------
+    statistics_ : 1-D array of length {n_features}
+        The 1-D array contains the mean of each feature calculated using
+        observed (i.e. non-missing) values.
```

Comment:
Note, because this is the only attribute, that these are only used in degenerate cases... (or note specifically which cases)